codestodians/food-basket-front-end
# Feat signup signout 153670662

#### What does this PR do?
This PR adds signup and signout functionality to the app.

#### Description of Task to be completed?
This PR includes
- seperation of signup into presentational and functional components.
- adding test setup that allows the use of enzyme to shallow render the components.
- adding a user reducer that updates the state on signup and signout.
- update tests and snapshopt for the signup component.
- adding auth actions(signup and signout).
- adding signout mutation.
- add redux saga for managing side effects.
- abstractions that were pulled from develop.
- toggle sign up and sign out buttons on the header.
- update the webpack config.
- add installed packages
- Add submission errors to the redux form(i.e display errors returned by the server on the form)

#### How should this be manually tested?
- Click the signup button on the modal
- Fill the signup form correctly and patiently wait for the form to submit.
- On submit, the sign out and sign in button on the modal change to the signout button(See screenshot)
- The auth state of the user is updated in redux store.
- Click sign out
- The sign out button is replaced by the sign up and sign in button(See screenshot)

#### Any background context you want to provide?
The sign out form did not perform any action on submit. This PR ensures that the user is stored in the database and redirects to the home page.

#### What are the relevant pivotal tracker stories?
[#153670662](https://www.pivotaltracker.com/story/show/153670662)

#### Screenshots (if appropriate)

##### *Header with signup and signin*
<img width="485" alt="screen shot 2017-12-21 at 11 05 31 pm" src="https://user-images.githubusercontent.com/20999363/34276961-17143062-e6a4-11e7-98aa-9c3f75808325.png">

##### *Header with signout only*
<img width="464" alt="screen shot 2017-12-21 at 11 07 02 pm" src="https://user-images.githubusercontent.com/20999363/34276970-1f56e2ec-e6a4-11e7-92b1-5d716dcb8683.png">

##### *Server side username validation*
<img width="696" alt="screen shot 2017-12-21 at 11 09 41 pm" src="https://user-images.githubusercontent.com/20999363/34276981-2d95e484-e6a4-11e7-8cc8-349efcc0eb12.png">

##### *Server side email validation*
<img width="696" alt="screen shot 2017-12-21 at 11 08 07 pm" src="https://user-images.githubusercontent.com/20999363/34276978-26b4961a-e6a4-11e7-873d-dd6aabed01f7.png">





#### Questions that may arise:
- Why not add the sign in functionality alongside?
> The PR as it stands,already has an enormous amount of files changed(over 20 already). The sign in functionality would only increase the files changed and may be better as another story.

- Why add redux saga?
> The use of redux saga had earlier being suggested by fisayo. My understanding of the need for redux saga( but is definitely not limited to) : 
> - dispatching an action when another action dispatches specifying relevant conditions e.g queuing dispatch of an action pending a timeout
> - handling concurrency problems e.g a user clicking on a button severally. redux saga can respond to the latest click
> - components are independent of action creators.
> - allows requests to be transformed before dispatching action

- Why are the files changed many?
>  This could have being better done.


#### Roadmap for the future
> Add a loading component to the form while the form is submitting.
